kadi/commands/commands_v2.py

```diff
@@ -0,0 +1,1185 @@
+            if prev_att is None:
+                # No previous attitude happens always for first manvr at the
+                # beginning of loads, and normally we just push on to the next
+                # OBS. But if we already have OBS command this is a problem.
+                if cmds_obs:
+                    logger.warning(f'WARNING: No previous attitude for {cmd["date"]}')
+                    log_context_obs(cmds, cmd)
+                prev_att = targ_att
+                continue
+            dur = manvr_duration(prev_att, targ_att)
+            params = {'manvr_start': cmd['date'],
+                      'prev_att': prev_att,
+                      'targ_att': targ_att,
+                      'npnt_enab': npnt_enab}
+            cmd_obs = {'idx': -1,
+                       'type': 'LOAD_EVENT',
+                       'tlmsid': 'OBS',
+                       'scs': 0,
+                       'step': 0,
+                       'source': cmd['source'],
+                       'vcdu': -1,
+                       'time': time + dur,
+                       'params': params,
+                       }
+            cmds_obs.append(cmd_obs)
+            prev_att = targ_att
+            targ_att = None
+
+    cmds_obs = CommandTable(rows=cmds_obs)
+    # NB: much faster to do this vectorized than when defining cmd_obs above.
+    cmds_obs.add_column(CxoTime(cmds_obs['time']).date, name='date', index=0)
+
+    # If an NSM occurs within a maneuver then remove that obs
+    nsms = cmds['tlmsid'] == 'AONSMSAF'
+    if np.any(nsms):
+        bad_idxs = []
+        for nsm_date in cmds['date'][nsms]:
+            for ii, cmd in enumerate(cmds_obs):
+                if nsm_date > cmd['params']['manvr_start'] and nsm_date <= cmd['date']:
+                    logger.info(f'NSM at {nsm_date} happened during maneuver preceding obs\n{cmd}')
+                    log_context_obs(cmds, cmd)
+                    bad_idxs.append(ii)
+        if bad_idxs:
+            logger.info(f'Removing obss at {bad_idxs}')
+            cmds_obs.remove_rows(bad_idxs)
+
+    # This is known to happen for cmds in AUG0103A, JAN2204B, JUL2604D due to
+    # non-load maneuvers that are not captured correctly. Run the v1-v2
+    # migration script to see. The prev_att is wrong and the incorrect maneuver
+    # timing results in overlaps. It should not happen for modern loads.
+    for cmd0, cmd1 in zip(cmds_obs[:-1], cmds_obs[1:]):
+        if cmd1['params']['manvr_start'] <= cmd0['date'] < cmd1['date']:
+            logger.warning(f'WARNING: fixing overlapping OBS cmds: \n{cmd0} \n{cmd1}')
+            date0 = CxoTime(cmd1['params']['manvr_start']) - 15 * u.s
+            cmd0['date'] = date0.date
+            cmd0['time'] = date0.secs
+
+    return cmds_obs
+
+
+# parameters from characteristics_Hardware for normal maneuvers
+MANVR_DELTA = 60.0
+MANVR_ALPHAMAX = 2.18166e-006  # AKA alpha
+MANVR_VMAX = 0.001309  # AKA omega
+
+
+def manvr_duration(q1, q2):
```

Comment:
Similar to the agasc comment, this probably doesn't belong here.  But not a blocker to beta promotion.

Reply:
Copy.